core/include/geometry/module_map.hpp

```diff
@@ -0,0 +1,392 @@
+             * starting at 2i + 2 instead.
+             */
+            if (e - (m + 1) > 0) {
+                z.emplace_back(2 * i + 2, m + 1, e);
+            }
+        }
+    }
+
+    /**
+     * @brief Special rounding function for binary trees.
+     *
+     * This function computes 2^n - 1, where n is the ceiling of log2 of the
+     * input plus one.
+     *
+     * This sounds like an odd operation, but consider that a binary tree (when
+     * complete) of depth k, can contain at most 2^k - 1 nodes. That is the
+     * number of spaces we need. However, our bottom level may have some empty
+     * slots, because it is rather unlikely that we will have exactly 2^k - 1
+     * nodes. Thus, we need to round up k from log2 of the input. However, we
+     * need to also keep in mind that 2^l (for some integer l) should be
+     * rounded up to 2^(l+1), not 2^l! Since we are computing l, this would
+     * mean we would be short one space in some edge cases.
+     *
+     * @param[in] i The number to round up.
+     *
+     * @return The specially rounded number.
+     */
+    static std::size_t round_up(std::size_t i) {
+        std::size_t n = 1;
+
+        while (n < i + 1) {
+            n *= 2;
+        }
+
+        return n - 1;
+    }
+
+    /**
+     * @brief Get a list of values and nodes from an existing std::map.
+     *
+     * Given an existing map, this function will attempt to find all the nodes,
+     * that is to say contiguous sub-ranges. It will also put the values into
+     * one big, contiguous array.
+     *
+     * @param[in] input An existing std::map module map.
+     *
+     * @return A pair of vectors, one has the values, and the other has the
+     * nodes.
+     */
+    static std::pair<std::vector<V>, std::vector<module_map_node>> node_list(
+        const std::map<K, V>& input) {
+        /*
+         * We start out by grabbing a list of all the keys in the map.
+         */
+        std::vector<K> keys;
+        keys.reserve(input.size());
+        for (const std::pair<K, V>& i : input) {
+            keys.push_back(i.first);
+        }
+
+        /*
+         * Next, we declare a vector of values, which is one of the two things
+         * that we will return in the end.
+         */
+        std::vector<V> values;
+        values.reserve(keys.size());
+
+        /*
+         * To find contiguous sub-ranges, the keys must be sorted. In general,
+         * it is likely that they will be pre-sorted.
+         */
+        std::sort(keys.begin(), keys.end());
+
+        /*
+         * This vector will hold our nodes. We'll update it as we go, and we
+         * start out with an incomplete first node, which represents the single
+         * value at the start of the key vector. We will add to this later!
+         */
+        std::vector<module_map_node> nodes;
+        nodes.push_back(module_map_node{
+            .start = keys.front(), .size = 1, .index = values.size()});
```

Comment:
Wait, isn't this C++20?

Reply:
Heh, you're right. I keep forgetting that this isn't widely accepted C++ yet, I use it in C code all the time because it's been valid there for a long time. :sweat_smile:

Good catch, I'll change it.

Reply:
Oh, I can't wait either :D